ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1212 @@
+ *                \       +-------------+
+ *                 \      |  TLV size   |   Generic TLV size
+ *                  \     +-------------+   (in bytes, 64-bit aligned)
+ *                   \    +-------------+
+ *                    \   | PLAT TLV sz |   Platform TLV size
+ *                     \--+-------------+   (in bytes, 64-bit aligned)
+ *
+ *                        +-------------+   Signature of the header, the trailer
+ *                        | Signature   |   and optionally the firmware image if
+ *                        +-------------+   a hash table is not stored.
+ *
+ *                        +-------------+
+ *                        |   Firmware  |
+ *                        |    image    |
+ *                        +-------------+
+ *                  ------+-------------+
+ *                 /      |+-----------+|
+ *                /       ||   TLV     ||   Information used to authenticate the
+ *  +-----------+/        ||           ||   firmware, stored in
+ *  |  Trailer  |         |+-----------+|   Type-Length-Value format.
+ *  +-----------+\        |+-----------+|
+ *                \       || Platform  ||   Specific platform information,
+ *                 \      ||   TLV     ||   stored in Type-Length-Value format.
+ *                  \     |+-----------+|
+ *                   -----+-------------+
+ *
+ *  TLV and platform TLV chunks:
+ *
+ *                        +-------------+
+ *                        |    Magic    |
+ *                        +-------------+
+ *                        +-------------+
+ *                        |    size     |  size of the TLV payload
+ *                        +-------------+
+ *                        +-------------+
+ *                        |     TLV     |  TLV structures
+ *                        |   payload   |
+ *                        +-------------+
+ */
+
+/* Firmware state */
+enum remoteproc_state {
+	REMOTEPROC_OFF,
+	REMOTEPROC_LOADED,
+	REMOTEPROC_STARTED,
+};
+
+#define RPROC_HDR_MAGIC		0x3543A468 /* random value */
+#define HEADER_VERSION		1
+
+/* Supported signature algorithm */
+enum remoteproc_sign_type {
+	RPROC_RSASSA_PKCS1_v1_5_SHA256 = 1,
+	RPROC_ECDSA_SHA256 = 2,
+};
+
+enum remoteproc_img_type {
+	REMOTEPROC_ELF_TYPE = 1,
+	REMOTEPROC_INVALID_TYPE = 0xFF
+};
+
+/* remoteproc_tlv structure offsets */
+#define RPROC_TLV_LENGTH_OF	U(0x02)
+#define RPROC_TLV_VALUE_OF	U(0x04)
+
+#define TLV_INFO_MAGIC		U(0x6907)
+#define PLAT_TLV_INFO_MAGIC	U(0x6908)
+
+/* TLV types */
+#define RPROC_TLV_SIGNTYPE	U(0x01)
+#define RPROC_TLV_HASHTYPE	U(0x02)
+#define RPROC_TLV_NUM_IMG	U(0x03)
+#define RPROC_TLV_IMGTYPE	U(0x04)
+#define RPROC_TLV_IMGSIZE	U(0x05)
+#define RPROC_TLV_HASHTABLE	U(0x10)
+#define RPROC_TLV_PKEYINFO	U(0x11)
+
+#define RPROC_TLV_SIGNTYPE_LGTH U(1)
+
+#define LE16_TO_CPU(x) (((x)[1] << 8) | (x)[0])
+#define LE32_TO_CPU(x) (((x)[3] << 24) | ((x)[2] << 16) | \
+			((x)[1] << 8)  | (x)[0])
```

Comment:
could use `TEE_U32_BSWAP()` and `TEE_U16_BSWAP()` from _utee_defines.h_

Reply:
If I am not mistaken, `TEE_U32_BSWAP`  allows to swap between little and big endian

Here we want to convert from Little endian to CPU endian( so LE or BE).
So using TEE_U32_BSWAP should not work.

To answer to such requirement we should need equivalent of the Linux helper:
https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/byteorder

Reply:
`TEE_U32_BSWAP()` and `TEE_U16_BSWAP()` do the very exact same thing as done here by `LE32_TO_CPU()` and `LE16_TO_CPU()`.